### ISSUE TITLE
fix(docs): resolve warnings logged by gulp public_docs task

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#2452]
cc @naomiblack 

TODO:

 - [ ] Fix broken links in `annotations.ts`: PR #2475 
 - [x] Fix checkAnchorLinks service in Dgeni to url decode normalized urls when checking dangling links: ~~https://github.com/angular/dgeni-packages/pull/119~~ 
 - [x] Update dgeni-packages dependency PR #2489
 - [x]  Fix document names/ids to not include full heritage of classes: #2462 
 - [x] Merge #2499 to get rid of Ambiguous Link warnings caused by #2486

Opening an issue to track work to fix errors when running the `gulp public_docs` task. @petebacondarwin tells me time is better spent addressing these issues than the ones generated by the `gulp docs` task. Based on current master (ddd5a235c39477970b8db8c37a1a63a82a389357), here are the warnings logged when running `gulp public_docs`:

```log
warn:    Invalid link (does not match any doc): "QueryList" - doc "DirectiveAnnotation extends  Injectable" (class)  - from file "angular2/src/core/annotations_impl/annotations.ts"
warn:    Invalid link (does not match any doc): "QueryList" - doc "DirectiveAnnotation extends  Injectable" (class)  - from file "angular2/src/core/annotations_impl/annotations.ts"
warn:    Invalid link (does not match any doc): "DynamicChangeDetection" - doc "ChangeDetection" (class)  - from file "angular2/src/change_detection/interfaces.ts"
warn:    Invalid link (does not match any doc): "JitChangeDetection" - doc "ChangeDetection" (class)  - from file "angular2/src/change_detection/interfaces.ts"
warn:    Invalid link (does not match any doc): "DynamicChangeDetection" - doc "ChangeDetection" (class)  - from file "angular2/src/change_detection/interfaces.ts"
warn:    Invalid link (does not match any doc): "JitChangeDetection" - doc "ChangeDetection" (class)  - from file "angular2/src/change_detection/interfaces.ts"
warn:    Invalid link (does not match any doc): "JitChangeDetection" - doc "DynamicChangeDetection extends  ChangeDetection" (class)  - from file "angular2/src/change_detection/change_detection.ts"
warn:    Invalid link (does not match any doc): "DynamicChangeDetection" - doc "JitChangeDetection extends  ChangeDetection" (class)  - from file "angular2/src/change_detection/change_detection.ts"
warn:    Invalid link (does not match any doc): "QueryList" - doc "QueryAnnotation extends  DependencyAnnotation" (class)  - from file "angular2/src/core/annotations_impl/di.ts"
warn:    Invalid link (does not match any doc): "QueryList" - doc "QueryAnnotation extends  DependencyAnnotation" (class)  - from file "angular2/src/core/annotations_impl/di.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "angular2/forms" (module)  - from file "angular2/forms.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "angular2/forms" (module)  - from file "angular2/forms.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "angular2/forms" (module)  - from file "angular2/forms.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "Control extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "ControlArray" - doc "Control extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "ControlGroup extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "ControlGroup extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "ControlArray" - doc "ControlGroup extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "ControlArray" - doc "ControlGroup extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "ControlArray extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "ControlArray extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "ControlArray extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "ControlArray extends  AbstractControl" (class)  - from file "angular2/src/forms/model.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "DefaultValueAccessor implements  ControlValueAccessor" (class)  - from file "angular2/src/forms/directives/default_value_accessor.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "SelectControlValueAccessor implements  ControlValueAccessor" (class)  - from file "angular2/src/forms/directives/select_control_value_acces
sor.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "FormBuilder" (class)  - from file "angular2/src/forms/form_builder.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "FormBuilder" (class)  - from file "angular2/src/forms/form_builder.ts"
warn:    Invalid link (does not match any doc): "ControlGroup" - doc "FormBuilder" (class)  - from file "angular2/src/forms/form_builder.ts"
warn:    Invalid link (does not match any doc): "Control" - doc "FormBuilder" (class)  - from file "angular2/src/forms/form_builder.ts"
warn:    Dangling Links Found in "partials/modules/angular2/annotations/index.html":
 - /angular2/annotations/ComponentAnnotation extends  Directive,
 - /angular2/annotations/DirectiveAnnotation extends  Injectable
warn:    Dangling Links Found in "partials/modules//angular2/annotations/DirectiveAnnotation extends  Injectable/index.html":
 - QueryList,
 - QueryList,
 - annotations/onChange,
 - annotations/onDestroy,
 - annotations/onAllChangesDone,
 - pipes/keyValDiff
warn:    Dangling Links Found in "partials/modules/angular2/change_detection/index.html":
 - /angular2/change_detection/ASTWithSource extends  AST,
 - /angular2/change_detection/AccessMember extends  AST,
 - /angular2/change_detection/LiteralArray extends  AST,
 - /angular2/change_detection/ImplicitReceiver extends  AST,
 - /angular2/change_detection/DehydratedException extends  BaseException,
 - /angular2/change_detection/ExpressionChangedAfterItHasBeenChecked extends  BaseException,
 - /angular2/change_detection/ChangeDetectionError extends  BaseException,
 - /angular2/change_detection/DynamicProtoChangeDetector extends  ProtoChangeDetector,
 - /angular2/change_detection/JitProtoChangeDetector extends  ProtoChangeDetector,
 - /angular2/change_detection/DynamicChangeDetector extends  AbstractChangeDetector,
 - /angular2/change_detection/NullPipe extends  Pipe,
 - /angular2/change_detection/NullPipeFactory extends  PipeFactory,
 - /angular2/change_detection/DynamicChangeDetection extends  ChangeDetection,
 - /angular2/change_detection/JitChangeDetection extends  ChangeDetection,
 - /angular2/change_detection/PreGeneratedChangeDetection extends  ChangeDetection
warn:    Dangling Links Found in "partials/modules//angular2/change_detection/ChangeDetection/index.html":
 - DynamicChangeDetection,
 - JitChangeDetection,
 - DynamicChangeDetection,
 - JitChangeDetection
warn:    Dangling Links Found in "partials/modules//angular2/change_detection/DynamicChangeDetection extends  ChangeDetection/index.html":
 - JitChangeDetection
warn:    Dangling Links Found in "partials/modules//angular2/change_detection/JitChangeDetection extends  ChangeDetection/index.html":
 - DynamicChangeDetection
warn:    Dangling Links Found in "partials/modules/angular2/core/index.html":
 - /angular2/core/SelfAnnotation extends  Visibility,
 - /angular2/core/AncestorAnnotation extends  Visibility,
 - /angular2/core/ParentAnnotation extends  Visibility,
 - /angular2/core/UnboundedAnnotation extends  Visibility,
 - /angular2/core/QueryAnnotation extends  DependencyAnnotation,
 - /angular2/core/AttributeAnnotation extends  DependencyAnnotation,
 - /angular2/core/QueryList<T> extends  BaseQueryList<T>
warn:    Dangling Links Found in "partials/modules//angular2/core/QueryAnnotation extends  DependencyAnnotation/index.html":
 - QueryList,
 - QueryList
warn:    Dangling Links Found in "partials/modules/angular2/di/index.html":
 - /angular2/di/NoBindingError extends  AbstractBindingError,
 - /angular2/di/AbstractBindingError extends  BaseException,
 - /angular2/di/AsyncBindingError extends  AbstractBindingError,
 - /angular2/di/CyclicDependencyError extends  AbstractBindingError,
 - /angular2/di/InstantiationError extends  AbstractBindingError,
 - /angular2/di/InvalidBindingError extends  BaseException,
 - /angular2/di/NoAnnotationError extends  BaseException
warn:    Dangling Links Found in "partials/modules/angular2/forms/index.html":
 - ControlGroup,
 - Control,
 - Control,
 - /angular2/forms/Control extends  AbstractControl,
 - /angular2/forms/ControlGroup extends  AbstractControl,
 - /angular2/forms/ControlArray extends  AbstractControl,
 - /angular2/forms/ControlNameDirective extends  ControlDirective,
 - /angular2/forms/FormControlDirective extends  ControlDirective,
 - /angular2/forms/NgModelDirective extends  ControlDirective,
 - /angular2/forms/ControlGroupDirective extends  ControlContainerDirective,
 - /angular2/forms/FormModelDirective extends  ControlContainerDirective implements  FormDirective,
 - /angular2/forms/TemplateDrivenFormDirective extends  ControlContainerDirective implements
    FormDirective,
 - /angular2/forms/DefaultValueAccessor implements  ControlValueAccessor,
 - /angular2/forms/CheckboxControlValueAccessor implements  ControlValueAccessor,
 - /angular2/forms/SelectControlValueAccessor implements  ControlValueAccessor
warn:    Dangling Links Found in "partials/modules//angular2/forms/Control extends  AbstractControl/index.html":
 - ControlGroup,
 - ControlArray
warn:    Dangling Links Found in "partials/modules//angular2/forms/ControlGroup extends  AbstractControl/index.html":
 - Control,
 - Control,
 - ControlArray,
 - ControlArray
warn:    Dangling Links Found in "partials/modules//angular2/forms/ControlArray extends  AbstractControl/index.html":
 - Control,
 - Control,
 - ControlGroup,
 - ControlGroup
warn:    Dangling Links Found in "partials/modules//angular2/forms/DefaultValueAccessor implements  ControlValueAccessor/index.html":
 - Control
warn:    Dangling Links Found in "partials/modules//angular2/forms/SelectControlValueAccessor implements  ControlValueAccessor/index.html":
 - Control
warn:    Dangling Links Found in "partials/modules//angular2/forms/FormBuilder/index.html":
 - ControlGroup,
 - Control,
 - ControlGroup,
 - Control
warn:    Dangling Links Found in "partials/modules/angular2/pipes/index.html":
 - /angular2/pipes/PromisePipe extends  Pipe,
 - /angular2/pipes/UpperCasePipe extends  Pipe,
 - /angular2/pipes/LowerCasePipe extends  Pipe,
 - /angular2/pipes/ObservablePipe extends  Pipe,
 - /angular2/pipes/JsonPipe extends  Pipe,
 - /angular2/pipes/IterableChanges extends  Pipe,
 - /angular2/pipes/KeyValueChanges extends  Pipe
warn:    83 'unmatched links'
```

**Comment by jeffbcross** [comment]
After diving a little deeper, I see what is happening, and I'm not sure the correct place to fix. The problem is that the keys in Dgeni's `aliasMap` sometimes contain spaces and inheritance/implementation descriptions, such as: `DynamicChangeDetector extends  AbstractChangeDetector`, but the links to these files usually just contain the class name. Even if a link did contain the full key name, the `@link` parser matches only up until the first space in the string, so it wouldn't be matched.

I think the right place to solve it is at the place where the keys are generated, to match the key up until a space or `<` character. What do you think, @petebacondarwin? If you can help point me to where this logic lives, I can submit a PR.

**Comment by petebacondarwin** [comment]
@jeffbcross I think the problem is caused here: https://github.com/angular/angular/blame/master/docs/dgeni-package/processors/readTypeScriptModules.js#L119-L140

@alexeagle added this block of code to help generate better `.d.ts` files but I think it needs a bit of love. Rather than modifying the name of the export, we should ensure that the `exportDoc` contains the necessary information and render it appropriately in the doc's template.

**Comment by jeffbcross** [comment]
@petebacondarwin would an appropriate solution be to add an additional alias with just the first part of the name here: https://github.com/angular/angular/blame/master/docs/dgeni-package/processors/readTypeScriptModules.js#L146

**Comment by jeffbcross** [comment]
The "Dangling Links" warnings are mostly related to links being url encoded, trying to match against keys that are decoded. https://github.com/angular/dgeni-packages/blob/39a713b502594c5f27a20f550351d5374a89e4cb/base/processors/checkAnchorLinks.js#L89

All but 1 warning (4 unmatched links for Injectable/index.html) go away if I change this line
`return allValidReferences[decodeURIComponent(normalizedLink + pathVariant)];`

@petebacondarwin is that a reasonable fix?

**Comment by petebacondarwin** [comment]
@jeffbcross - the change to `checkAnchorLinks` sounds reasonable.

**Comment by jeffbcross** [comment]
I opened a PR on `dgeni-packages` with the fix for `checkAnchorLinks`: https://github.com/angular/dgeni-packages/pull/119

I'll submit a PR to angular/angular after that lands to update the dgeni-packages dependency.

(And to recap an offline convo I had with @petebacondarwin : Pete is going to take a stab at a PR in the next day or so to properly address the "Invalid Links" issue that this PR addresses)

**Comment by jeffbcross** [comment]
@petebacondarwin I opened another PR to deal with a side effect of your fix to `readTypeScriptModules`. If duplicate aliases are added, a bunch of "Ambiguous Link" warnings show up.

So once my three open PRs are LGTM'd and merged, I think we'll be down to zero warnings for the `docs/angular.io` task.

**Comment by jeffbcross** [comment]
The only remaining issue to be merged is the fixes to self-links inside of annotations.ts, which I'm not sure the correct way to annotate. The only way that works without warnings is to specify the link as a relative path: `{@link /angular2/annotations/onChange}`. Waiting on advisement from @petebacondarwin  on #2475 .

**Comment by petebacondarwin** [comment]
I think that this final set of commits: 992293a196761def8720464cd8e5d43394ab2876, 14d28d74730c92b0ff849c08d5e9a1de89d93178 and dd23bab3adc27f2453b578d00ecaa4723e2375a8 resolve all the warnings.
